chips/rp2040/src/pio_spi.rs

```diff
@@ -0,0 +1,562 @@
+// Licensed under the Apache License, Version 2.0 or the MIT License.
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// Copyright Tock Contributors 2025.
+//
+// Author: Jason Hu <jasonhu2026@u.northwestern.edu>
+//         Anthony Alvarez <anthonyalvarez2026@u.northwestern.edu>
+
+//! Programmable Input Output (PIO) hardware test file.
+use crate::clocks::{self};
+use crate::pio::{Pio, PioRxClient, PioTxClient, SMNumber, StateMachineConfiguration};
+use core::cell::Cell;
+use kernel::deferred_call::{DeferredCall, DeferredCallClient};
+use kernel::hil::spi::cs::{ChipSelectPolar, Polarity};
+use kernel::hil::spi::SpiMasterClient;
+use kernel::hil::spi::{ClockPhase, ClockPolarity};
+use kernel::utilities::cells::MapCell;
+use kernel::utilities::cells::OptionalCell;
+use kernel::utilities::leasable_buffer::SubSliceMut;
+use kernel::{hil, ErrorCode};
+
+// Since auto push / pull is set to 8 for the purposes of writing in bytes
+// rather than words, values read in have to be bitshifted by 24
+const AUTOPULL_SHIFT: usize = 24;
+
+// Frequency of system clock, for rate changes
+const SYSCLOCK_FREQ: u32 = 125_000_000;
+
+// Leading edge clock phase + Idle low clock
```

Comment:
Is this (and other pio assembly) from scratch or taken from the `pico-examples` repo or other reference implementations? If the former, it would be good to document what each of these are actually doing for the reader unfamiliar with pio assembly. If the latter, at least a link or other reference to the original would help track provenance in case of bugs etc arising in the future.